fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneRecordCursor.java

```diff
@@ -288,16 +214,52 @@ private synchronized IndexReader getIndexReader() throws IOException {
         return FDBDirectoryManager.getManager(state).getIndexReader(groupingKey);
     }
 
-    private void performScan() throws IOException {
+    private void maybePerformScan() throws IOException {
+        if (lookupResults != null && currentPosition + skip < pageSize) {
+            return;
+        }
+        if (lookupResults != null) {
+            lookupResults.close();
+        }
+
+        TopDocs newTopDocs = searchForTopDocs(Math.min(limitRemaining + skip, pageSize));
+
+        lookupResults = RecordCursor.fromIterator(executor, Arrays.stream(newTopDocs.scoreDocs).iterator()).skip(skip)
```

Comment:
Doesn't something need to clear `skip` for subsequent pages?

Reply:
I know we decided not to attempt to further refactoring to parallelize across pages, but I will point out that having a paged cursor of docs and then skipping that and then mapping that to `IndexEntry` might make it easier to get the `skip` logic correct.

Reply:
Not it has one final ```skip``` and one mutable ```leftToSkip``` which tracks the left of records to skip. it is cleaned up after skipping all records now. And there are new added test to verify this.